### PR TITLE
Implement DataTree.isel and DataTree.sel

### PR DESCRIPTION
**Pull request by shoyer**
These methods handle dimensions that are not missing on some nodes properly, as discussed in https://github.com/pydata/xarray/issues/8949 

Adding additional indexing methods (e.g., `head`, `tail`, `thin`, `interp`, `reindex`), should be pretty easy to do in this style.

- [x] Tests added
- [x] New functions/methods are listed in `api.rst`

### CODE REVIEW

File: xarray/core/datatree.py [hunk 1]
```diff
@@ -1607,3 +1614,61 @@ def to_zarr(
...
+            all_dims.update(node._node_dims)
+        indexers = drop_dims_from_indexers(indexers, all_dims, missing_dims)
+
+        result = {}
+        for node in self.subtree:
+            node_indexers = {k: v for k, v in indexers.items() if k in node.dims}
+            node_result = func(node.dataset, node_indexers)
+            for k in node_indexers:
+                if k not in node._node_coord_variables and k in node_result.coords:
+                    del node_result.coords[k]
```
Comment (TomNicholas) [inline xarray/core/datatree.py:1635]: So this is basically saying "if the result of indexing would cause an inherited coordinate to be duplicated, then remove that duplicated coordinate from the result"?

Doesn't this imply an inefficiency, in that we are potentially doing the work to index coordinates that we're only going to drop afterwards?

File: xarray/core/datatree.py [hunk 2]
```diff
@@ -1607,3 +1614,61 @@ def to_zarr(
             compute=compute,
             **kwargs,
         )
+
+    def _selective_indexing(
+        self,
+        func: Callable[[Dataset, Mapping[Any, Any]], Dataset],
+        indexers: Mapping[Any, Any],
+        missing_dims: ErrorOptionsWithWarn = "raise",
+    ) -> Self:
```
Comment (TomNicholas) [inline xarray/core/datatree.py:1623]: Just a note for devs
```suggestion
    ) -> Self:
    """
    Apply an indexing operation over the subtree, handling missing dimensions and inherited coordinates gracefully by only applying indexing at each node selectively.
    """
```

File: xarray/core/datatree.py [hunk 3]
```diff
@@ -1607,3 +1614,61 @@ def to_zarr(
...
+        return type(self).from_dict(result, name=self.name)
+
+    def isel(
+        self,
+        indexers: Mapping[Any, Any] | None = None,
+        drop: bool = False,
+        missing_dims: ErrorOptionsWithWarn = "raise",
+        **indexers_kwargs: Any,
+    ) -> Self:
+        """Positional indexing."""
```
Comment (TomNicholas) [inline xarray/core/datatree.py:1646]: These docstrings will be public-facing, so we should at least copy the detailed version from `Dataset.isel` (for now maybe just remove the examples and change the `See Also`).

File: xarray/core/datatree.py [hunk 4]
```diff
@@ -1607,3 +1614,61 @@ def to_zarr(
...
+        method: str | None = None,
+        tolerance: int | float | Iterable[int | float] | None = None,
+        drop: bool = False,
+        **indexers_kwargs: Any,
+    ) -> Self:
+        """Label-based indexing."""
+
+        def apply_indexers(dataset, node_indexers):
+            # TODO: reimplement in terms of map_index_queries(), to avoid
+            # redundant index look-ups on child nodes
```
Comment (TomNicholas) [inline xarray/core/datatree.py:1668]: Is this related to my comment above or a separate issue?

File: xarray/tests/test_datatree.py [hunk 5]
```diff
@@ -1561,26 +1560,86 @@ def test_filter(self):
...
+        )
+
+        expected = DataTree.from_dict(
+            {
+                "/": xr.Dataset(coords={"x": 2}),
+                "/child": xr.Dataset({"foo": 4}),
+            }
+        )
+        actual = tree.isel(x=-1)
+        assert_equal(actual, expected)
```
Comment (TomNicholas) [inline xarray/tests/test_datatree.py:1612]: When this passes the `x` coordinate gets duplicated doesn't it? As the result is a non-indexed scalar. Does that mean this would have failed if you had used the definition of `assert_identical` from https://github.com/pydata/xarray/pull/9473 to check for duplication?

Comment (TomNicholas) [reply xarray/tests/test_datatree.py:1612]: IIUC this should also pass

```python
actual = tree.isel(x=[0])
assert_identical(actual, expected)
```

because the result will have inherited instead of duplicated coordinates instead.

File: xarray/core/datatree.py [hunk 6]
```diff
@@ -1607,3 +1614,61 @@ def to_zarr(
...
+    def _selective_indexing(
+        self,
+        func: Callable[[Dataset, Mapping[Any, Any]], Dataset],
+        indexers: Mapping[Any, Any],
+        missing_dims: ErrorOptionsWithWarn = "raise",
+    ) -> Self:
+        all_dims = set()
+        for node in self.subtree:
+            all_dims.update(node._node_dims)
+        indexers = drop_dims_from_indexers(indexers, all_dims, missing_dims)
```
Comment (TomNicholas) [inline xarray/core/datatree.py:1627]: This is neat.

File: xarray/core/datatree.py (hunk 2 above)
Comment (shoyer) [reply xarray/core/datatree.py:1623]: done

File: xarray/core/datatree.py (hunk 1 above)
Comment (shoyer) [reply xarray/core/datatree.py]: I added a comment to explain. It is indeed a small source of inefficiency.

File: xarray/core/datatree.py (hunk 3 above)
Comment (shoyer) [reply xarray/core/datatree.py]: Done

File: xarray/core/datatree.py (hunk 4 above)
Comment (shoyer) [reply xarray/core/datatree.py]: It's a separate issue.

File: xarray/tests/test_datatree.py (hunk 5 above)
Comment (shoyer) [reply xarray/tests/test_datatree.py:1612]: I wrote special logic for removing duplicated scalar coordinates to handle just this!

There is no longer a need to implement a different version of inheritance for `assert_identical` than for `assert_equal`, because we no longer allow any cases in the DataTree data model where inheritance is ambiguous:
- Deduplication of coordinates with an index removes them from child nodes.
- Other coordinates are no longer inherited.

**Review by TomNicholas — APPROVED** [review]
Thanks for the explanations! The comments definitely make it easier to understand the subtleties.
